### PR TITLE
Postgres integration + experiments

### LINKED ISSUE #499: Evaluate alternatives to MariaDB (and RDBMS)
1. Answer following questions.
      - [ ] Estimate number of articles in the `articles` table + number of rows in the `sentences` table (neuroscience articles only).
             _Note_: If this number is large, then MariaDB is not suitable.
      - [x] Analyze which actions (`INSERT`, `SELECT`, `UPDATE`, ...) we are expected performing on the `sentences` table. 
      E.g. do we need to filter by xxx ... (is it a `FULLTEXT` search? is it a filter on a indexed/indexable column)?
            _Note_: We may want to consider NoSQL alternatives to the current RDBMS architecture.
2. Discuss and complete migration from MariaDB to Postgres.
      - [x] Make changes in the code.
      - [ ] Complete migration to new Postgres database.

### LINKED ISSUE #527: PostreSQL migration
Following #499 , we now want to migrate to PostgreSQL.

### PR DESCRIPTION
**Pull request by jankrepl**
This PR adds the `postgres` backend to the `init` and `add` entrypoints. It also adds it to the integration test `tests/integration/test_bbs_database.py`. 

Closes #499 
Fixes #527.

## Description
Why is the CI failing? There is this `PytestUnraisableExceptionWarning` ([docs](https://docs.pytest.org/en/6.2.x/reference.html#pytest.PytestUnraisableExceptionWarning)). There are several ways how to handle this:
- Figure out the core reason why it is raised and fix it
- Ignore it locally with `@pytest.mark.filterwarnings("ignore::pytest.PytestUnraisableExceptionWarning")` - already done elsewhere in our sourc ecode
- Ignore it globally with `pytest -p no:unraisableexception`
- Abandon this test / change it so that no exception occurs (e.g. trying a different driver than `pg8000`?)

## How to test?
```
pytest -p no:unraisableexception tests/integration/test_bbs_database.py 
```

## Checklist

- [ ] This PR refers to an issue from the [issue tracker](https://github.com/BlueBrain/Search/issues).
  (if it is not the case, please create an issue first).
- [ ] Unit tests added.
  (if needed)
- [ ] Documentation and `whatsnew.rst` updated.
  (if needed)
- [ ] `setup.py` and `requirements.txt` updated with new dependencies.
  (if needed)
- [ ] Type annotations added.
  (if a function is added or modified)
- [ ] All CI tests pass.

### CODE REVIEW

**Comment by Stannislav** [comment]
I'd go with the option of using `@pytest.mark.filterwarnings("ignore::pytest.PytestUnraisableExceptionWarning")`

**Review by FrancescoCasalegno — COMMENTED** [review]
Thank you for doing this @jankrepl, looks great!

Concerning `PytestUnraisableExceptionWarning`, I couldn't really understand where it is coming from by reading the error message below. I only have the impression that it is happening during `setup_backend` when `request.param` (from the parametrization `params`) is `postgresql`.

<details>
<summary>Error message</summary>

  ```python
  
  ________________ ERROR at setup of test_bbs_database[postgres] _________________
  
  cls = <class '_pytest.runner.CallInfo'>
  func = <function call_runtest_hook.<locals>.<lambda> at 0x7f725a499f70>
  when = 'setup'
  reraise = (<class '_pytest.outcomes.Exit'>, <class 'KeyboardInterrupt'>)
  
      @classmethod
      def from_call(
          cls,
          func: "Callable[[], TResult]",
          when: "Literal['collect', 'setup', 'call', 'teardown']",
          reraise: Optional[
              Union[Type[BaseException], Tuple[Type[BaseException], ...]]
          ] = None,
      ) -> "CallInfo[TResult]":
          excinfo = None
          start = timing.time()
          precise_start = timing.perf_counter()
          try:
  >           result: Optional[TResult] = func()
  
  .tox/py39/lib/python3.9/site-packages/_pytest/runner.py:311: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  .tox/py39/lib/python3.9/site-packages/_pytest/runner.py:255: in <lambda>
      lambda: ihook(item=item, **kwds), when=when, reraise=reraise
  .tox/py39/lib/python3.9/site-packages/pluggy/_hooks.py:265: in __call__
      return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  .tox/py39/lib/python3.9/site-packages/pluggy/_manager.py:80: in _hookexec
      return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  .tox/py39/lib/python3.9/site-packages/_pytest/unraisableexception.py:83: in pytest_runtest_setup
      yield from unraisable_exception_runtest_hook()
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
      def unraisable_exception_runtest_hook() -> Generator[None, None, None]:
          with catch_unraisable_exception() as cm:
              yield
              if cm.unraisable:
                  if cm.unraisable.err_msg is not None:
                      err_msg = cm.unraisable.err_msg
                  else:
                      err_msg = "Exception ignored in"
                  msg = f"{err_msg}: {cm.unraisable.object!r}\n\n"
                  msg += "".join(
                      traceback.format_exception(
                          cm.unraisable.exc_type,
                          cm.unraisable.exc_value,
                          cm.unraisable.exc_traceback,
                      )
                  )
  >               warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
  E               pytest.PytestUnraisableExceptionWarning: Exception ignored in: <socket.socket fd=-1, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6>
  E               
  E               Traceback (most recent call last):
  E                 File "/home/runner/work/Search/Search/.tox/py39/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 661, in __connect
  E                   self.dbapi_connection = connection = pool._invoke_creator(self)
  E               ResourceWarning: unclosed <socket.socket fd=22, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('127.0.0.1', 57760)>
  
  .tox/py39/lib/python3.9/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning
  ```  
</details>

Do you have any more clear idea of what's going on?

As for the solution I agree with @Stannislav that we can use `@pytest.mark.filterwarning`.

**Comment by FrancescoCasalegno** [comment]
By the way, this PR is marked as "partially" closing https://github.com/BlueBrain/Search/issues/499.

But is there really anything actionable left in #499? Can't we mark this PR as closing #499?
You also created on Confluence the page `Blue Brain Search (BBS) Home > BBS Text Mining Tools > PostgreSQL` so I would say we have everything ready to start `bbs_database add` resources to the new `PostgresSQL` database.

In conclusion, unless I missed any other point, personally I don't think it is necessary to wait for having populated the whole new DB before closing #499 and I think we can close it with this PR.

What do you think @jankrepl?

File: setup.py [hunk 1]
```diff
@@ -62,6 +62,7 @@
     "pandas>=1",
     # Serialization framework on top of dataclasses, e.g. 'Article' to and from JSON.
     "mashumaro",
+    "pg8000",
```
[the remark this answers is not in the document]
Comment (FrancescoCasalegno) [reply setup.py:65]: `pg8000` is fine with me, I just wanted to understand why not `psycopg2`.

I will resolve this thread.

[the remark this answers is not in the document]
Comment (jankrepl) [reply setup.py:67]: Done in 9737360bd1dc0a4b2f742cfa71f76fdc747cbc93

**Review by Stannislav — APPROVED** [review]
thanks :)
